src/relay/backend/contrib/vsi_npu/codegen.cc

```diff
@@ -0,0 +1,425 @@
+#include "op_map/op_setup.h"
+
+#include <tvm/relay/attrs/image.h>
+#include <tvm/relay/attrs/nn.h>
+#include <tvm/relay/attrs/reduce.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/transform.h>
+#include <tvm/relay/type.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/registry.h>
+
+#include <fstream>
+#include <iostream>
+#include <numeric>
+#include <cassert>
+#include <sstream>
+
+#include "tim/transform/layout_inference.h"
+
+namespace tvx = tim::vx;
+
+namespace tvm {
+namespace relay {
+namespace contrib {
+namespace vsi_npu {
+
+using TensorInfoTable = std::map<Expr, std::vector<tim::vx::TensorSpec>>;
+
+void quant_info_infer(VxOpTable& op_tb, Expr now_expr, bool is_input) {
+  auto now_opsetup = op_tb[now_expr];
+  Expr pre_expr;
+  if ((now_opsetup->pCallbackexpr_ == nullptr ||
+      now_opsetup->pCallbackexpr_->ptr_pre_callback_ == nullptr) && is_input
+      ) {
+    return;
+  } else if((now_opsetup->pCallbackexpr_ == nullptr ||
+      now_opsetup->pCallbackexpr_->ptr_pre_callback_ == nullptr
+      || op_tb[now_expr]->specs_[0].quantization_.ZeroPoints().size() == 0)&& !is_input ){
+      return;
+  } else {
+    pre_expr = now_opsetup->pCallbackexpr_->ptr_pre_callback_->expr_;
+  }
+
+  auto pre_opsetup = op_tb[pre_expr];
+  auto ptr_callback = pre_opsetup->pCallbackexpr_;
+
+  if (now_opsetup->specs_[0].datatype_ == tvx::DataType::FLOAT32 ||
+      pre_opsetup->specs_[0].datatype_ == tvx::DataType::FLOAT32 ||
+      now_opsetup->specs_[0].datatype_ == tvx::DataType::BOOL8 ||
+      pre_opsetup->specs_[0].datatype_ == tvx::DataType::BOOL8) {
+    return;
+  }
+
+  tvx::Quantization& now_quant_info = now_opsetup->specs_[0].quantization_;
+
+  std::vector<int32_t> zps;
+  std::vector<float> scales;
+  if (now_quant_info.Type() == tvx::QuantType::NONE) {
+    zps = {0};
+    scales = {1.0};
+    now_quant_info.SetType(tvx::QuantType::ASYMMETRIC).SetScales({1.0}).SetZeroPoints({0});
+  } else {
+    zps = now_quant_info.ZeroPoints();
+    scales = now_quant_info.Scales();
+  }
+
+  while (ptr_callback &&
+         op_tb[ptr_callback->expr_]->specs_[0].quantization_.ZeroPoints().size() == 0) {
+    Expr expr = ptr_callback->expr_;
+    auto datatype = GetTvxType(expr->checked_type().as<TensorTypeNode>()->dtype);
+    if (datatype != tim::vx::DataType::INT32) {
+      op_tb[expr]
+          ->specs_[0]
+          .quantization_.SetType(tvx::QuantType::ASYMMETRIC)
+          .SetScales(scales)
+          .SetZeroPoints(zps);
+    }
+    ptr_callback = ptr_callback->ptr_pre_callback_;
+  }
+}
+
+template <typename T, typename T2>
+void attribute_transform(const T &attrs, T2 &attrs_num) {
+
+  std::transform(attrs.begin(), attrs.end(), attrs_num.begin(),
+                 [](const PrimExpr &attrs_num) {
+                   return static_cast<uint32_t>(
+                       attrs_num.as<IntImmNode>()->value);
+                 });
+};
+
+std::shared_ptr<tvx::Tensor> createVxOPerand(TensorInfoTable tensor_info,
```

Comment:
can you follow the C++ style guide here https://google.github.io/styleguide/cppguide.html#Function_Names? e.g.`CreateVxOperand`

Reply:
Will fix.